src/main/scala/mesosphere/marathon/upgrade/ReadinessBehavior.scala

```diff
@@ -79,26 +82,44 @@ trait ReadinessBehavior { this: Actor with ActorLogging =>
     * The #taskIsReady function is called, when the task is ready according to the app definition.
     */
   //scalastyle:off cyclomatic.complexity
-  def readinessBehavior: Receive = {
-    def taskIsRunning(taskFn: Task.Id => Unit): Receive = {
-      case MesosStatusUpdateEvent(slaveId, taskId, "TASK_RUNNING", _, `appId`, _, _, _, `versionString`, _, _) =>
-        taskFn(taskId)
-    }
-
-    def taskIsHealthy(taskFn: Task.Id => Unit): Receive = {
-      case HealthStatusChanged(`appId`, taskId, `version`, true, _, _) if !healthy(taskId) => taskFn(taskId)
+  val readinessBehavior: Receive = {
+
+    def taskRunBehavior: Receive = {
+      def runningTaskWithoutReady(taskId: Task.Id): Unit = {
+        log.debug(s"Started task is ready: $taskId")
+        healthy += taskId
+        ready += taskId
+        taskStatusChanged(taskId)
+      }
+      def runningTaskWithReady(taskId: Task.Id): Unit = {
+        healthy += taskId
+        initiateReadinessCheck(taskId)
+      }
+      def taskIsRunning(taskFn: Task.Id => Unit): Receive = {
+        case MesosStatusUpdateEvent(slaveId, taskId, "TASK_RUNNING", _, `appId`, _, _, _, `versionString`, _, _) =>
+          taskFn(taskId)
+      }
+      taskIsRunning(if (app.readinessChecks.isEmpty) runningTaskWithoutReady else runningTaskWithReady)
     }
 
-    def startedTaskIsReady(taskId: Task.Id): Unit = {
-      log.debug(s"Started task is ready: $taskId")
-      healthy += taskId
-      ready += taskId
-      taskIsReady(taskId)
+    def taskHealthBehavior: Receive = {
+      def initiateReadinessOnRun: Receive = {
+        case MesosStatusUpdateEvent(_, taskId, "TASK_RUNNING", _, `appId`, _, _, _, `versionString`, _, _) =>
+          initiateReadinessCheck(taskId)
+      }
+      def taskIsHealthy: Receive = {
+        case HealthStatusChanged(`appId`, taskId, `version`, true, _, _) if !healthy(taskId) =>
+          log.info(s"Task $taskId now healthy for app ${app.id.toString}")
+          healthy += taskId
+          if (app.readinessChecks.isEmpty) ready += taskId
+          taskStatusChanged(taskId)
+      }
+      val onStarted = if (app.readinessChecks.nonEmpty) initiateReadinessOnRun else Actor.emptyBehavior
```

Comment:
To make it consistent with the line 102, I'd also call this `taskIsRunning`.

Although I actually prefer something like `handleTaskRunning` and `handleTaskHealthy`.